include/xsimd/types/xsimd_avx_shuffling.hpp

```diff
@@ -0,0 +1,148 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille and Sylvain Corlay                     *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XSIMD_AVX_SHUFFLING_HPP
+#define XSIMD_AVX_SHUFFLING_HPP
+
+#include "xsimd/types/xsimd_avx_conversion.hpp"
+
+namespace xsimd 
+{
+    // Version of select(a, b) with compile time mask
+    template <class B>
+    struct selector_immediate;
+
+    template <class T, std::size_t N>
+    struct selector_immediate<batch<T, N>>
+    {
+        using type = int32_t;
+    };
+
+    template <typename selector_immediate<batch<float, 8>>::type P, class T>
+    batch<T, 8> select(const batch<T, 8>& a, const batch<T, 8>& b);
+
+    template <typename selector_immediate<batch<double, 4>>::type P, class T>
+    batch<T, 4> select(const batch<T, 4>& a, const batch<T, 4>& b);
+
+    constexpr static int AaBbCcDd = 1;
+    constexpr static int AaCc = 2;
+    constexpr static int BbDd = 3;
+    constexpr static int AaBbEeFf = 4;
+    constexpr static int ABabEFef = 5;
+    constexpr static int CcDdGgHh = 6;
+    constexpr static int CDcdGHgh = 7;
```

Comment:
If these possibilities are mutually exclusive, I would make this an enum (class).

From a high-level design point of view, however, I am a bit uneasy with spelling out all the possibilities like this, due to the combinatorics involved. Could it scale all the way to AVX-512's vectors? To possible future extensions to smaller data types (Intel's SIMD instruction set supports 16-bit ints, the theoretical minimum is 8-bit)? To future wider instructions sets like ARM SVE?

Maybe it would be better to wait for precise use cases, and until then only implement obvious stuff like ({a, b, c, d}, {A, B, C, D}) -> ({a, A, b, B}, {c, C, d, D}). Studying the various SIMD instruction sets to see which interleavings are commonly supported could also be a good idea.

Reply:
yep, i already made it an enum in my version, haven't pushed yet! 

yes, i also agree that it might get cluttered .. therefore I was thinking maybe we can use this as sort of "patterns" that are to be used in combination with the bitsize of the interleave.

so e.g. doing an AaCc shuffle on a 8 bit width would result in 

AaBbCcDdEeFfGg_PpQqRrSsTtUuVvWw (I hope I spelled it out correctly).

The other option would be to find meaningful function names (e.g. in MIPP they have 

https://github.com/aff3ct/MIPP/blob/f0aabd427b9557124e7c8d384f4383a8c4f1f5cf/src/mipp_impl_AVX.hxx

- interleavelo2
- interleavehi2
- interleave
- interleavehi
- interleavelo

of all the different types.